src/third_party/dartino/two_space_heap.h

```diff
@@ -136,17 +132,27 @@ class TwoSpaceHeap {
 // Helper class for copying HeapObjects.
 class ScavengeVisitor : public RootCallback {
  public:
-  explicit ScavengeVisitor(Program* program, TwoSpaceHeap* heap)
+  explicit ScavengeVisitor(Program* program, TwoSpaceHeap* heap, Chunk* to_chunk)
       : program_(program),
-        to_start_(heap->unused_semi_space_->single_chunk_start()),
-        to_size_(heap->unused_semi_space_->single_chunk_size()),
+        to_start_(to_chunk->start()),
+        to_size_(to_chunk->size()),
         from_start_(heap->space()->single_chunk_start()),
         from_size_(heap->space()->single_chunk_size()),
-        to_(heap->unused_semi_space_),
+        to_(program, to_chunk),
         old_(heap->old_space()),
         record_(&dummy_record_),
         water_mark_(heap->water_mark_) {}
 
+  SemiSpace* to_space() { return &to_; }
+
+  void complete_scavenge() {
+    bool work_found = true;
+    while (work_found) {
+      work_found = to_.complete_scavenge(this);
+      work_found |= old_->complete_scavenge(this);
```

Comment:
It is a little subtle that this wouldn't work with ||= even when you're using bools. Maybe it would be simpler to just write:

```
if (old_->complete_scavenge(this)) work_found = true;
```